### PR TITLE
doc: VM metrics documentation

### PR DESCRIPTION
**Pull request by fromanirh**
Related-To: https://github.com/kubevirt/kubevirt/issues/296
Signed-off-by: Francesco Romani <fromani@redhat.com>

### CODE REVIEW

**Comment by fromanirh** [comment]
Proposal about how to address and solve https://github.com/kubevirt/kubevirt/issues/296

File: docs/vm-metrics.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# Overview
+VM expose a different, and partially overlapping, set of metrics with
+respect to containers. This document explain the differences and how
+those metrics are made available to the cluster, to be consumed using
+the standard kubernetes facilities.
```
Comment (fabiand) [inline docs/vm-metrics.md]: capital K, Kubernetes

File: docs/vm-metrics.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+## libvirt metrics
+[libvirt](https://libvirt.org/) is the virtualization framework that
+kubevirt uses; in order to describe the VM metrics in kubevirt, the
```
Comment (fabiand) [inline docs/vm-metrics.md]: capital K and V, KubeVirt

File: docs/vm-metrics.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+first step is to inspect what libvirt provides, and how it collect the
+data.
+
+In a nutshell, libvirt uses both the QEMU configuration and the standard
+facilites of the linux kernel to do the resource limitation and
+accounting. As prime example, libvirt uses cgroups, to do resource
```
Comment (fabiand) [inline docs/vm-metrics.md]: As -> A?

File: docs/vm-metrics.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+accounting for memory, cpu usage, I/O. Thus, we can consider this subset
+of metrics to fully overlap with what monitoring system (e.g.
+[cAdvisor](https://github.com/google/cadvisor)) already provides.
+Let's call this set `system metrics`.
```
Comment (fabiand) [inline docs/vm-metrics.md]: We need to be careful here.

cgroups are used in both cases, tho there are different processes in the livirt cgroups and in the kubelet cgoups.

Not sure if it matters.

Comment (fabiand) [inline docs/vm-metrics.md]: _system metrics_ is not code, what about using italics?

File: docs/vm-metrics.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+Libvirt exposes a set of hypervisor and operation-specific metrics.
+Examples are the migration statistics. Those metrics are usually
+available only while the operation is in progress, or shortly after.
+Let's call this set `hypervisor metrics`
```
Comment (fabiand) [inline docs/vm-metrics.md]: Same here, italics?

File: docs/vm-metrics.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+Lastly, VMs often run a Guest Agent, a deamon which periodically
+collects and reports back to the hypervisor another set of metrics.
+Let's call this set `guest metrics`
```
Comment (fabiand) [inline docs/vm-metrics.md]: Same here, italics?

File: docs/vm-metrics.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+The last two sets have no counterpart in the containers ecosystem, but
```
Comment (fabiand) [inline docs/vm-metrics.md]: Both could actually be seen as application metrics in the Kube world, but yes, on the Kube level there is no coutnerpart.

File: docs/vm-metrics.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+| guest state reason                 | hypervisor    | -         | -           |
+| memory balloon current             | hypervisor    | -         | -           |
+| memory balloon maximum             | hypervisor    | -         | -           |
+| virtual CPU count                  | hypervisor    | -         | -           |
+| virtual CPU maximum                | hypervisor    | -         | -           |
+| virtual CPU #x state               | hypervisor    | -         | -           |
+| virtual CPU #x total time          | hypervisor    | -         | -           |
+| network interface count            | hypervisor    | -         | -           |
+| block device count                 | hypervisor    | -         | -           |
+| migration downtime                 | hypervisor    | yes       |             |
```
Comment (fabiand) [inline docs/vm-metrics.md]: Any reason for not continuing to use a `-` in The _provided by_ column?

File: docs/vm-metrics.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+**NOTE** the list of guest agent provided metrics is partial, because
+a lot of different metrics can be collected for the guest. The presented
+entries are a subset of what the [QEMU Guest Agent](https://wiki.qemu.org/Features/GuestAgent)
+provides, which is used by virtualization managers like
+[oVirt](http://www.ovirt.org)
+
+From now on, the overlapping metrics  (aka 'system metrics') will be ignored,
+and we will focus on the 'hypervisor' and 'guest agent' metrics.
+
+## collection of VM metrics
```
Comment (fabiand) [inline docs/vm-metrics.md]: capital C, Collection?

File: docs/vm-metrics.md [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+A natural fit for the kubevirt specific metrics is the [custom metrics](
+https://github.com/kubernetes/kubernetes/blob/release-1.2/docs/proposals/custom-metrics.md).
+All the metrics are accessible through the libvirt API.
+
+The collection of the VM metrics will be done by a separate pod, implemented as kubernetes DaemonSet.
```
Comment (fabiand) [inline docs/vm-metrics.md]: capital K, Kubernetes

File: docs/vm-metrics.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+4. Keep the common prefixes among related metrics as long as possible. Examples
+
+- **BAD**: `VM0.migration.processedMemory`, `VM0.migration.remainingMemory`
+
+- **GOOD**: `VM0.migration.memoryProcessed`, `VM0.migration.memoryRemaining`
+
+Once name is exported, it is considered part of the API stability and treated accordingly.
+
+TODO: what about name clashes? at least report them? We can do that?
```
Comment (fabiand) [inline docs/vm-metrics.md]: We'd need somebody to find clashes.

File: docs/vm-metrics.md [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+to set the names:
+
+1. A name should be in the form `{VM name}.{group}.{metric name}`. The dot (`.`) is used as separator.
+   the `camelCase` should be used for composite words. Example: `VM0.migration.totalData`
+
+2. Abbreviations should be used only if their meaning is well known and not ambiguous. Examples:
+
+- **BAD**:
+  `VM0.migration.bps`
+  What `bps` is? Is it Bytes Per Second or Bits per Second? Or something else more specific?
```
Comment (fabiand) [inline docs/vm-metrics.md]: Yes, eventually encoding the unit in the name might make sense.

File: docs/vm-metrics.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+A natural fit for the kubevirt specific metrics is the [custom metrics](
+https://github.com/kubernetes/kubernetes/blob/release-1.2/docs/proposals/custom-metrics.md).
+All the metrics are accessible through the libvirt API.
+
+The collection of the VM metrics will be done by a separate pod, implemented as kubernetes DaemonSet.
+This is because the metrics are available from a centralized source, the libvirt daemon.
+To increase resilience, each node should expose all the metrics it knows about, avoiding
+a single point of failure per cluster. The resulting architecture is shown in the figure below.
+
+![VM metrics diagram](vm-metrics.png "VM Metrics")
```
Comment (fromanirh) [inline docs/vm-metrics.md]: wrong diagram, I (hastly) copied another one and got it wrong. Libvirt and vAdvisor must run on each node. Will update it ASAP.

**Comment by fabiand** [comment]
I'd appreciate that we start looking into providing additional components i.e. to showcase this metrics - i.e. add an optional graphana manifest

File: docs/vm-metrics.md (hunk 4 above)
Comment (smarterclayton) [reply docs/vm-metrics.md]: This is being called core metrics or resource metrics in kubernetes

File: docs/vm-metrics.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+provides, which is used by virtualization managers like
+[oVirt](http://www.ovirt.org)
+
+From now on, the overlapping metrics  (aka 'system metrics') will be ignored,
+and we will focus on the 'hypervisor' and 'guest agent' metrics.
+
+## collection of VM metrics
+A natural fit for the KubeVirt specific metrics is the [custom metrics](
+https://github.com/Kubernetes/Kubernetes/blob/release-1.2/docs/proposals/custom-metrics.md).
+All the metrics are accessible through the libvirt API.
```
Comment (smarterclayton) [inline docs/vm-metrics.md]: There can only be one custom metrics provider per cluster for now, which is going to be a prometheus gathering high value metrics), so keep that in mind

File: docs/vm-metrics.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+To have the full set the VM metrics, a guest agent like the QEMU Guest Agent should be run
+in the virtual machines. This step, albeit recommended, is totally left up to the administrator
+of the cluster.
+
+
+## VM metrics naming rules
+
+In order to minimize the clashes and make the VM metrics name predictable, we adopt the following rules
+to set the names:
```
Comment (smarterclayton) [inline docs/vm-metrics.md]: Are you trying to name these to be consistent with Kube?  The standard format for Kubernetes is prometheus, so I would expect you to follow the conventions there for metrics that will be gathered (use labels for unique indentificstion of entities, for example).

File: docs/vm-metrics.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+- The provider provide time series of data without holes in the sequence
+  (see normalization notes above).
+
+- The provider will never attempt to recover any past data.
+
+
+## Implementation notes
+
+To implement the VM metrics collection and reporting, we will initially
+adapt the [vAdvisor project](https://github.com/KubeVirt/vAdvisor),
```
Comment (smarterclayton) [inline docs/vm-metrics.md]: You're talking about the agent on the node collection?  Or something higher level?

File: docs/vm-metrics.md (hunk 3 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: fixed

File: docs/vm-metrics.md (hunk 4 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: Not sure either, but the concept I was try to convey is that the underlying monitoring mechanism is the same, so we can expect the same metric to be delivered one way or the other; in other words, there is nothing that libvirt reports that cAdvisor can't report as well -and vice versa.
Anyway I'll keep this point in mind and will check in the future.

Comment (fromanirh) [reply docs/vm-metrics.md]: OK about italics. Thanks @smarterclayton for pointing this out, will check kubernetes doc and will align the naming and the usage of terms.

File: docs/vm-metrics.md (hunk 5 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: will fix.

File: docs/vm-metrics.md (hunk 6 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: will fix.

File: docs/vm-metrics.md (hunk 8 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: It was a placeholder. I was looking for the equivalent concept in kubernetes, initial research was unfruitful. Now I learned those are 'core metrics', so I can drop the column.

File: docs/vm-metrics.md (hunk 9 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: Will fix.

File: docs/vm-metrics.md (hunk 11 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: Ok, so there is no automated way to detect them IIUC

File: docs/vm-metrics.md (hunk 14 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: OK, thanks for pointing this out

File: docs/vm-metrics.md (hunk 15 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: I was looking for feedback about this. Is the prometheus convention good enough for us? (I'm not hinting yes or no, just asking)

File: docs/vm-metrics.md (hunk 16 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: I mean the agent on the node, to report metrics we care about and cAdvisor doesn't know

File: docs/vm-metrics.md (hunk 14 above)
Comment (fromanirh) [reply docs/vm-metrics.md]: Relevant (AFAIU) documentation: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md

Comment (fromanirh) [reply docs/vm-metrics.md]: @smarterclayton could you please point me to some k8s documentation about this limitation of only one custom metrics provider? I want to learn more (e.g. why there is this constraint, if/we it is planned to be lifted, if any other constraint exist) but I didn't find complete documentation yet. Thanks!

File: docs/vm-metrics.md (hunk 12 above)
Comment (MarSik) [reply docs/vm-metrics.md]: bps is definitely bits per seconds btw :) Byte uses uppercase B when the unit is written properly. But it should not have its own leaf, unless you want to expose multiple different scales.

File: docs/vm-metrics.md (hunk 15 above)
Comment (MarSik) [reply docs/vm-metrics.md]: Prometheus format is less strict than what you proposed, because you can have multiple different orthogonal labels instead of just one fixed metrics name.

File: docs/vm-metrics.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+
+## Consuming the VM metrics
+
+In order to make the task of the consumer easier, the VM metrics provider pod normalizes the
+exported dataset in the following manner
+
+- It always export all the metrics it knows about, even if the real data is not yet available.
+  Such metrics report the special value [TODO: which one? does this
+  concept exist in Kubernetes?] until the real data is available.
```
Comment (MarSik) [inline docs/vm-metrics.md]: It might be possible to report the last modification date too, Prometheus format supports timestamps.

**Comment by fromanirh** [comment]
ping?

File: docs/vm-metrics.md [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+
+![VM metrics monitoring pipeline diagram](vm-metrics.png "VM Metrics monitoring pipeline")
+
+To have the full set the VM metrics, a guest agent like the QEMU Guest Agent should be run
+in the virtual machines. This step, albeit recommended, is totally left up to the administrator
+of the cluster.
+
+The key components of the monitoring pipeline are:
+
+- The KubeVirt clustr agent: we can use a standard Prometheus server instance
```
Comment (stu-gott) [inline docs/vm-metrics.md]: typo? clustr->cluster

File: docs/vm-metrics.md [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+
+- **BAD**: `VM0-migration-processedMemory`, `VM0-migration-remainingMemory`
+
+- **GOOD**: `VM0-migration-memoryProcessed`, `VM0-migration-memoryRemaining`
+
+Once name is exported, it is considered part of the API stability and treated accordingly.
+
+## Final diagram
+
+The following diagram illustrates how the KyubeVirt monitoring pipeline will look like
```
Comment (stu-gott) [inline docs/vm-metrics.md]: Typo: KyubeVirt

File: docs/vm-metrics.md [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+- avoid or at least minimize the clashes: wherever useful, it it should be possible to
+  disaggregate the metrics per-VM. For example, one could be interested in the metrics related to
+  the migration of one specific VM, to make sure the migration is converging and to take actions
+  if it is not.
+
+In order to minimize the clashes and make the VM metrics name predictable, we adopt the following rules
+to set the names:
+
+1. **TO BE CHECKED** If namespaces are available, they should be used: each VM has its own namespace - the VM name or UUID,
+   to be decided. Otherwise, a metric name should include the VM identifier (name or UUID) as prefix.
```
Comment (rmohr) [inline docs/vm-metrics.md]: In case of Prometheus that would be labels, not a metric prefix, right?

File: docs/vm-metrics.md [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+[libvirt](https://libvirt.org/) is the virtualization framework that
+KubeVirt uses. The libvirt control daemon, libvirtd, does its own independent
+collection of the VM metrics. 
+libvirtd uses both the QEMU configuration and the standard
+facilites of the linux kernel to do the resource limitation and
+accounting. For example, libvirt uses cgroups, to do resource
+accounting for memory, cpu usage, I/O.
+
+Libvirtd provides metrics that roughly overlaps with Kubernetes
+[core metrics](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md)
```
Comment (rmohr) [inline docs/vm-metrics.md]: Looks like in that proposal, the core metrics should not be extensible by design, so we can't add core metrics which might be relevant for VMs only there?

File: docs/vm-metrics.md [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+  For example, the VM metric $VM_migration_data_total will be always
+  reported even before the first migration was attempted, if ever.
+
+- The node agent will have one option to keep the last known values of ephemeral metrics
+  after libvirt no longer reports them, for example again the $VM_migration_data_total metric
+  will be reported even after the migration is completed
+
+The normalization features will be controlled by node agent options (e.g. command line switches).
+
+### Metric naming rules
```
Comment (rmohr) [inline docs/vm-metrics.md]: That looks good.

File: docs/vm-metrics.md [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+should require the minimal amount of resources, both as runtime and as container image.
+We know from the experience that shared storage issues may cause VMs to hang, and libvirt
+may in turn hang while collecting metrics from stuck VMs. We want the Node Agent to cope
+gracefully with this scenario, and keep reporting the metrics of the live VMs, skipping
+the stuck VMs. Reporting of stuck VMs is part of health monitoring, which is out of the
+scope of this document.
+
+The options to implement the KubeVirt node agent are:
+
+#### vAdvisor
```
Comment (rmohr) [inline docs/vm-metrics.md]: If we really only export prometheus metrics, it might not even be worth to temporarily use vAdvisor. Although I thought it might make sense in the past. Collecting and exposing prometheus metrics is extremely simple in go.

File: docs/vm-metrics.md [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+- No scalability issues
+- Better integration with the ecosystem
+- Minimal image size
+
+**CONS**
+- We will pay the rewrite cost upfront. But if we keep the design basics, we can minimize the effort.
+  [Initial experimentations](https://github.com/fromanirh/go-vAdvisor) gave promising results about
+  the development speed
+
+#### collectd
```
Comment (rmohr) [inline docs/vm-metrics.md]: I would try to focus on the mostly used tools which are afaik heapster, prometheus and cAdvisor. snapd and collectd seem to be mentioned quite regularly too. I don't have enough insight in real world deployments to see if snapd and collectd are important enough right now. Anyone?

File: docs/vm-metrics.md [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+
+**CONS**
+- Reporting of ephemeral metrics could need work and acceptance upstream
+- Collectd uses one [implicit and hardcoded mapping](https://github.com/collectd/collectd/blob/master/src/write_prometheus.c#L405)
+  for prometheus labels which is a poor fit for the KubeVirt use case.
+- long release cycle
+
+## KubeVIRT API Adapters: VM metrics naming rules
+
+The task of the KubeVirt HPA API adapter is to expose the metrics collected by the cluster agent
```
Comment (rmohr) [inline docs/vm-metrics.md]: If HPA stands for Horizontal Pod Autoscaler, I am a little bit confused. HPA is for sure a consumer of Heapster/prometheus metrics, but how is that related to this proposal? Maybe I misunderstand HPA? Or are you thinking about scaling VMs based on that?

File: docs/vm-metrics.md [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+- **GOOD**: `VM0-migration-memoryProcessed`, `VM0-migration-memoryRemaining`
+
+Once name is exported, it is considered part of the API stability and treated accordingly.
+
+## Final diagram
+
+The following diagram illustrates how the KyubeVirt monitoring pipeline will look like
+with all the components described in this document deployed
+
+![VM metrics monitoring pipeline example](vm-metrics-example.png "VM Metrics monitoring pipeline example")
```
Comment (rmohr) [inline docs/vm-metrics.md]: Could you add existing component names (if they exist) to the boxes? E.g. "Monitoring CLuster Agent" probably relates to Heapster or Prometheus?

**Comment by fabiand** [comment]
Thanks for the work.

I'm closing this PR for now, we can revive it once this becomes relevant again.

**Comment by yanirq** [comment]
[[epic-detailed-monitoring] Prepare components for metrics collection for scale tests](https://trello.com/c/qe7PGx2c/8-epic-detailed-monitoring-prepare-components-for-metrics-collection-for-scale-tests)
